File: data-structures-and-algorithms/binary-trees/easy/cousins-in-binary-tree/approach1/cousinsInBinaryTree.js
function isCousins(root, x, y) {
  // Save the depth of the first node.
  recordedDepth = null
  isCousin = false

  function dfs(node, depth, x, y) {
    if (node == null) return false

    // Don't go beyond the depth restricted by the first node found.
    if ((recordedDepth && depth) > recordedDepth) return false

    if ((node.val == x) || (node.val == y)) {
      if (recordedDepth == null) {
        // Save depth for the first node.
        recordedDepth = depth
      }

      // Return true, if the second node is found at the same depth
      return recordedDepth == depth
    }

    let left = dfs(node.left, depth + 1, x, y)
    let right = dfs(node.right, depth + 1, x, y)

    // recordedDepth != depth + 1 would ensure node x and y are not immediate
    // child nodes, otherwise they would become siblings.
    if ((left && right) && (recordedDepth != depth + 1)) {
      isCousin = true
    }

    return left || right
  }

  dfs(root, 0, x, y)
  return isCousin
}

File: data-structures-and-algorithms/binary-trees/easy/cousins-in-binary-tree/approach1/cousins_in_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        # Save the depth of the first node.
        self.recorded_depth = None
        self.is_cousin = False

    def dfs(self, node, depth, x, y):
        if node is None:
            return False

        # Don't go beyone the depth restricted by the first node found.
        if self.recorded_depth and depth > self.recorded_depth:
            return False

        if (node.val == x) or (node.val == y):
            if self.recorded_depth == None:
                # Save depth for the first node.
                self.recorded_depth = depth

            # Return true, if the second node is found at the same depth.
            return self.recorded_depth == depth

        left = self.dfs(node.left, depth + 1, x, y)
        right = self.dfs(node.right, depth + 1, x, y)

        # recordedDepth != depth + 1 would ensure node x and y are not immediate
        # child nodes, otherwise they would become siblings.
        if (left and right) and (self.recorded_depth != depth + 1):
            self.is_cousin = True

        return left or right

    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.dfs(root, 0, x, y)
        return self.is_cousin
